truffle/src/main/java/org/jruby/truffle/core/string/StringNodes.java

```diff
@@ -1355,6 +1365,307 @@ private int prevCharHead(Encoding enc, byte[]bytes, int p, int s, int end) {
+
+            if (cr == CR_7BIT || cr == CR_VALID)
+                return nil();
+
+            enc = rope.getEncoding();
+            if (repl != nil()) {
+                repl = strCompatAndValidNode.call(frame, string, "str_compat_and_valid", repl, rubyEncoding);
+                tainted |= isTaintedNode.executeIsTainted(repl);
+            }
+
+            if (enc.isDummy()) {
+                return nil();
+            }
+            encidx = enc;
+
+            if (enc.isAsciiCompatible()) {
+                byte[] pBytes = value.unsafeBytes();
+                int p = value.begin();
+                int e = p + value.getRealSize();
+                int p1 = p;
+                boolean rep7bit_p;
+                if (block != nil()) {
+                    repBytes = null;
+                    rep = 0;
+                    replen = 0;
+                    rep7bit_p = false;
+                }
+                else if (repl != nil()) {
+                    final Rope replRope = rope((DynamicObject) repl);
+                    final ByteList replValue = RopeOperations.getByteListReadOnly(replRope);
+                    repBytes = replValue.unsafeBytes();
+                    rep = replValue.begin();
+                    replen = replValue.getRealSize();
+                    rep7bit_p = (replRope.getCodeRange() == CodeRange.CR_7BIT);
+                }
+                else if (encidx == UTF8Encoding.INSTANCE) {
+                    repBytes = SCRUB_REPL_UTF8;
+                    rep = 0;
+                    replen = repBytes.length;
+                    rep7bit_p = false;
+                }
+                else {
+                    repBytes = SCRUB_REPL_ASCII;
+                    rep = 0;
+                    replen = repBytes.length;
+                    rep7bit_p = false;
+                }
+                cr = CR_7BIT;
+
+                p = StringSupport.searchNonAscii(pBytes, p, e);
+                if (p == -1) {
+                    p = e;
+                }
+                while (p < e) {
+                    int ret = enc.length(pBytes, p, e);
+                    if (MBCLEN_NEEDMORE_P(ret)) {
+                        break;
+                    }
+                    else if (MBCLEN_CHARFOUND_P(ret)) {
+                        cr = CR_VALID;
+                        p += MBCLEN_CHARFOUND_LEN(ret);
+                    }
+                    else if (MBCLEN_INVALID_P(ret)) {
+                    /*
+                     * p1~p: valid ascii/multibyte chars
+                     * p ~e: invalid bytes + unknown bytes
+                     */
+                        int clen = enc.maxLength();
+                        if(buf == null){
+                            buf = RopeConstants.EMPTY_ASCII_8BIT_ROPE;
+                        }
+                        if (p > p1) {
+                               buf = makeConcatNode.executeMake(buf, RopeOperations.ropeFromByteList(new ByteList(pBytes, p1, p - p1)), enc);
```

Comment:
I don't think you need to create a `ByteList` here. You should be able to get by with `ArrayUtils.extractRange` and `RopeOperations.create`.